### DISCUSSION TITLE
AttributeError: module 'awkward' has no attribute 'layout'

### DISCUSSION THREAD

**Discussion by picassopotato** [scikit-hep/awkward#660] [Q&A]
Hi, sorry if this is a very basic question but.. I'm trying to get things to work and I'm following the https://hsf-training.github.io/hsf-training-uproot-webpage/ guide. The only context is that I have a file with a TTree containing several TBranches. I want to make a dict holding the branches and their arrays, but I get this:

```>>> import uproot
>>> f = uproot.open("uproot-tutorial-file.root")
>>> branches = f["Events"].arrays()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/uproot/behaviors/TBranch.py", line 1120, in arrays
    arrays,
  File "/usr/local/lib/python3.7/site-packages/uproot/behaviors/TBranch.py", line 3430, in _ranges_or_baskets_to_arrays
    uproot.source.futures.delayed_raise(*obj)
  File "/usr/local/lib/python3.7/site-packages/uproot/source/futures.py", line 46, in delayed_raise
    raise exception_value.with_traceback(traceback)
  File "/usr/local/lib/python3.7/site-packages/uproot/behaviors/TBranch.py", line 3407, in basket_to_array
    branch,
  File "/usr/local/lib/python3.7/site-packages/uproot/interpretation/numerical.py", line 115, in final_array
    output = library.finalize(output, branch, self, entry_start, entry_stop)
  File "/usr/local/lib/python3.7/site-packages/uproot/interpretation/library.py", line 483, in finalize
    if isinstance(array, awkward.layout.Content):
AttributeError: module 'awkward' has no attribute 'layout'
```

To get here I just did pip install uproot. I have uproot 4.0.1 and awkward 0.14.0. Would you please lend a hand?

**Comment by jpivarski** [comment]
The problem is mixing Uproot 4 with Awkward 0. Update pip packages with

```bash
pip uninstall awkward awkward0 awkward1 uproot uproot3 uproot4
```

enough times that you get rid of all existing versions, then

```bash
pip install awkward uproot
```

and if you _need_ old interfaces, you can also install

```bash
pip install awkward0 uproot3    # the old ones
```

You can use old and new in the same Python process, but be sure your versions are all to the right of the name-change transition line:

![](https://user-images.githubusercontent.com/1852447/105192214-7c166400-5afd-11eb-94be-0ae9df79e611.png)

**Comment by picassopotato** [comment]
Oh sorry, no idea why I had awkward0. Or, awkward but in version 0? But thank you that worked, and thanks for the swift repsonse!

**Comment by jpivarski** [comment]
For more context, this is part of a major Awkward+Uproot upgrade that completed in December 2020. The error you saw was because names were switched, and the names were switched because some analyses might be depending on the old interfaces. We want to allow gradual adoption by the ability to import both old and new in the same Python process, which wouldn't be possible if the version number was switched.

Unfortunately for pip, you have to upgrade packages manually. If something brings in an old version of some package as a dependency, it won't get upgraded unless some package explicitly requires a new version or you manually upgrade it. Technically (i.e. from pip's perspective), Uproot 4 does not require Awkward Array, so that it is more portable, so that's why you didn't get it automatically.

![image](https://user-images.githubusercontent.com/1852447/105201235-41fd9000-5b06-11eb-83c8-7e2ee19f85c4.png)

**Comment by picassopotato** [comment]
"If something brings in an old version of some package as a dependency" yes this turned out to be the issue, installing uproot-methods caused awkward to downgrade. I've seen that vector (https://vector.readthedocs.io/en/latest/index.html) is the recommended way to handle Lorentz vectors, but I'm a little bit lost. I have a TTree with TLorentzVectors and I just need to get things like .Pt() and .Phi(). Do I need to recreate those functions from fX fY fZ and fE? Do you have a good guide for this sort of thing? Thanks!

**Reply by picassopotato** [reply]
@jpivarski sorry to be a bother but may I kindly remind you of this?

**Reply by jpivarski** [reply]
Actually, it's been starred in my inbox for days. It prompted me to talk with the author of Vector about trading some of our responsibilities. I'll be working on the Vector package personally to get it into shape for analysis. My work on that will start in March and will probably take about a month.

For the short-term, writing functions for simple things like `pt` and `phi` is an option (see https://github.com/scikit-hep/uproot3-methods/blob/master/uproot3_methods/classes/TLorentzVector.py).

Another option is to use Coffea's vector classes: https://coffeateam.github.io/coffea/modules/coffea.nanoevents.methods.vector.html?highlight=vector

**Reply by picassopotato** [reply]
Awesome, thanks!
